src/sys/timerfd.rs

```diff
@@ -0,0 +1,124 @@
+//! Timer API via file descriptors.
+//!
+//! Timer FD is a Linux-only API to create timers and get expiration notifications through file
+//! descriptors.
+//!
+//! For more documentation, please read [timerfd_create(2)](http://man7.org/linux/man-pages/man2/timerfd_create.2.html).
+//!
+//! # Examples
+//!
+//! Create a new one-shot timer that expires after 5 seconds.
+//! ```no_run
+//! # use nix::sys::timerfd::{TimerFd, ClockId, TimerFlags, TimerSetTimeFlags, Expiration};
+//! # use nix::sys::time::{TimeSpec, TimeValLike};
+//! #
+//! // We create a new monotonic timer.
+//! let fd = TimerFd::new(ClockId::Monotonic, TimerFlags::empty()).unwrap();
+//!
+//! // We set a new one-shot timer in 5 seconds.
+//! fd.settime(TimerSetTimeFlags::empty(), Expiration::OneShot(TimeSpec::seconds(5))).unwrap();
+//! ```
+use crate::errno::Errno;
+use crate::sys::time::TimeSpec;
+use crate::Result;
+use bitflags::bitflags;
+use libc::c_int;
+use std::os::unix::io::{AsRawFd, RawFd};
+
+#[derive(Debug, Clone, Copy)]
+pub struct TimerFd {
+    fd: i32,
+}
+
+impl AsRawFd for TimerFd {
+    fn as_raw_fd(&self) -> RawFd {
+        self.fd
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ClockId {
+    Realtime,
+    Monotonic,
+    BootTime,
+    RealtimeAlarm,
+    BootTimeAlarm,
+}
+
+libc_bitflags! {
+    pub struct TimerFlags: c_int {
+        TFD_NONBLOCK;
+        TFD_CLOEXEC;
+    }
+}
+
+bitflags! {
+    pub struct TimerSetTimeFlags: libc::c_int {
+        const TFD_TIMER_ABSTIME = libc::TFD_TIMER_ABSTIME;
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TimerSpec(libc::itimerspec);
+
+impl AsRef<libc::itimerspec> for TimerSpec {
+    fn as_ref(&self) -> &libc::itimerspec {
+        &self.0
+    }
+}
+
+impl From<Expiration> for TimerSpec {
+    fn from(expiration: Expiration) -> TimerSpec {
+        match expiration {
+            Expiration::OneShot(t) => TimerSpec(libc::itimerspec {
+                it_interval: libc::timespec {
+                    tv_sec: 0,
+                    tv_nsec: 0,
+                },
+                it_value: *t.as_ref(),
+            }),
+            Expiration::IntervalDelayed(start, interval) => TimerSpec(libc::itimerspec {
+                it_interval: *interval.as_ref(),
+                it_value: *start.as_ref(),
+            }),
+            Expiration::Interval(t) => TimerSpec(libc::itimerspec {
+                it_interval: *t.as_ref(),
+                it_value: *t.as_ref(),
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum Expiration {
+    OneShot(TimeSpec),
+    IntervalDelayed(TimeSpec, TimeSpec),
+    Interval(TimeSpec),
+}
+
+impl TimerFd {
+    pub fn new(clockid: ClockId, flags: TimerFlags) -> Result<Self> {
+        let clid = match clockid {
```

Comment:
I think you can skip this block if you define ClockId using the `libc_enum` macro.

Reply:
Ok refactored ClockId to use libc_enum.